options/glibc/generic/execinfo.cpp

```diff
@@ -1,11 +1,67 @@
 #include <execinfo.h>
+#include <unwind.h>
 
 #include <bits/ensure.h>
 #include <mlibc/debug.hpp>
 
-int backtrace(void **, int) {
-	mlibc::infoLogger() << "mlibc: backtrace() is a stub!" << frg::endlog;
-	return 0;
+namespace {
+
+frg::optional<void *> libgccHandle = frg::null_opt;
+_Unwind_Reason_Code (*unwindBacktrace)(_Unwind_Trace_Fn, void *) = nullptr;
+_Unwind_Ptr (*unwindGetIP)(_Unwind_Context *) = nullptr;
+
+struct UnwindState {
+	void **frames;
+	int count;
+	int current_frame = 0;
+};
+
+_Unwind_Reason_Code trace(_Unwind_Context *context, void *arg) {
+	uintptr_t ip = unwindGetIP(context);
+
+	if (ip) {
+#if defined(__x86_64__) || defined(__i386__)
+		ip--;
+#elif defined(__aarch64__) || defined(__loongarch64)
+		ip -= 4;
+#elif defined(__riscv) || defined(__m68k__)
+		ip -= 2;
+#else
+#warning "Missing support for architecture"
+		ip--;
+#endif
+	}
+
+	UnwindState *state = static_cast<UnwindState *>(arg);
+	state->frames[state->current_frame++] = reinterpret_cast<void *>(ip);
+	return (state->current_frame >= state->count) ? _URC_END_OF_STACK : _URC_NO_REASON;
```

Comment:
Doesn't it make more sense to check for overflow before writing? For example, this will fail for `state->count == 0`.